kernel/kernel-api/src/main/java/io/delta/kernel/types/StructField.java

```diff
@@ -48,31 +50,96 @@ public class StructField {
           METADATA_ROW_INDEX_COLUMN_NAME,
           LongType.LONG,
           false,
-          FieldMetadata.builder().putBoolean(IS_METADATA_COLUMN_KEY, true).build());
+          FieldMetadata.builder().putBoolean(IS_METADATA_COLUMN_KEY, true).build(),
+          Collections.emptyList());
 
   public static final String COLLATIONS_METADATA_KEY = "__COLLATIONS";
 
   ////////////////////////////////////////////////////////////////////////////////
   // Instance Fields / Methods
   ////////////////////////////////////////////////////////////////////////////////
 
+  /**
+   * Represents a type change for a field, containing the original and new primitive types.
+   *
+   * <p>Type changes are actually persisted in metadata attached to StructFields but the rules for
+   * where the metadata is attached depend on if the change is for nested arrays/maps or primitive
+   * types.
+   */
+  public static class TypeChange {
+    private final BasePrimitiveType from;
+    private final BasePrimitiveType to;
+
+    public TypeChange(BasePrimitiveType from, BasePrimitiveType to) {
+      this.from = Objects.requireNonNull(from, "from type cannot be null");
+      this.to = Objects.requireNonNull(to, "to type cannot be null");
+    }
+
+    public BasePrimitiveType getFrom() {
+      return from;
+    }
+
+    public BasePrimitiveType getTo() {
+      return to;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      TypeChange that = (TypeChange) o;
+      return Objects.equals(from, that.from) && Objects.equals(to, that.to);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(from, to);
+    }
+
+    @Override
+    public String toString() {
+      return String.format("TypeChange(from=%s,to=%s)", from, to);
+    }
+  }
+
   private final String name;
   private final DataType dataType;
   private final boolean nullable;
   private final FieldMetadata metadata;
+  private final List<TypeChange> typeChanges;
 
   public StructField(String name, DataType dataType, boolean nullable) {
     this(name, dataType, nullable, FieldMetadata.empty());
   }
 
   public StructField(String name, DataType dataType, boolean nullable, FieldMetadata metadata) {
+    this(name, dataType, nullable, metadata, Collections.emptyList());
+  }
+
+  public StructField(
+      String name,
+      DataType dataType,
+      boolean nullable,
+      FieldMetadata metadata,
+      List<TypeChange> typeChanges) {
```

Comment:
hmm.. we need a builder for StructField soon. (not related to this PR)

Reply:
yep :)